EpikCord/client/http.py

```diff
@@ -0,0 +1,433 @@
+        """
+        logger.info(f"Waiting for bucket {self.hash} to be set.")
+        await self.event.wait()
+        logger.info(f"Done waiting for bucket {self.hash}.")
+
+    def set(self):
+        """
+        Sets the bucket.
+        """
+        logger.info(f"Setting bucket {self.hash}.")
+        self.event.set()
+
+    def clear(self):
+        """
+        Clears the bucket.
+        """
+        logger.info(f"Clearing bucket {self.hash}.")
+        self.event.clear()
+
+    def __eq__(self, other: Bucket):
+        if isinstance(other, Bucket):
+            return self.hash == other.hash
+        return False
+
+
+class TopLevelBucket:
+    def __init__(self, *, major_parameters: Dict[str, Any]):
+        """
+        Parameters
+        ----------
+        major_parameters: Dict[str, Any]
+            The major parameters for this Bucket
+
+        Attributes
+        ----------
+        major_parameters: Dict[str, Any]
+            The major parameters for this Bucket
+        """
+        self.event = asyncio.Event()
+        self.event.set()
+        self.major_parameters: Dict[str, Any] = major_parameters
+
+    async def wait(self):
+        """
+        Waits for the bucket to be set.
+        """
+        logger.info(f"Waiting for bucket {self.major_parameters} to be set.")
+        await self.event.wait()
+        logger.info(f"Done waiting for bucket {self.major_parameters}.")
+
+    def set(self):
+        """
+        Sets the bucket.
+        """
+        logger.info(f"Setting bucket {self.major_parameters}.")
+        self.event.set()
+
+    def clear(self):
+        """
+        Clears the bucket.
+        """
+        logger.info(f"Clearing bucket {self.major_parameters}.")
+        self.event.clear()
+
+    def __eq__(self, other: TopLevelBucket):
+        if isinstance(other, TopLevelBucket):
+            return self.major_parameters == other.major_parameters
+        return False
+
+    def __str__(self):
+        return f"channel_id={self.major_parameters.get('channel_id')}, guild_id={self.major_parameters.get('guild_id')}, webhook_id={self.major_parameters.get('webhook_id')}, webhook_token={self.major_parameters.get('webhook_token')}"
+
+
+class HTTPClient:
+    error_mapping: Dict[
+        int,
+        Union[Type[NotFound], Type[Forbidden], Type[Unauthorized], Type[BadRequest]],
+    ] = {400: BadRequest, 401: Unauthorized, 403: Forbidden, 404: NotFound}
+
+    def __init__(self, token: str, *, version: int = 10):
+        self.token: str = token
+        self.version: int = version
+        self.session: aiohttp.ClientSession = aiohttp.ClientSession(
+            headers={
+                "Authorization": f"Bot {self.token}",
+                "User-Agent": f"DiscordBot (https://github.com/EpikCord/EpikCord.py {__version__})",
+            },
+            json_serialize=lambda x, *__, **___: json.dumps(x).decode("utf-8")  # type: ignore
+            if _ORJSON
+            else json.dumps(x),
```

Comment:
This is messy and will be needed for the gateway, so you should move this code into a utility function.